monai/transforms/utils.py

```diff
@@ -2509,5 +2511,25 @@ def distance_transform_edt(
     return convert_data_type(r_vals[0] if len(r_vals) == 1 else r_vals, output_type=type(img), device=device)[0]
 
 
+def apply_affine_to_points(data: torch.Tensor, affine: torch.Tensor, dtype: DtypeLike | torch.dtype):
+    """
+    apply affine transformation to a set of points.
+
+    Args:
+        data: input data to apply affine transformation, should be a tensor of shape (C, N, 2 or 3).
+        affine: affine matrix to be applied, should be a tensor of shape (3, 3) or (4, 4).
+        dtype: output data dtype.
+    """
+    data_: torch.Tensor = convert_to_tensor(data, track_meta=False, dtype=torch.float64)
+    affine = to_affine_nd(data_.shape[-1], affine)
+
+    homogeneous = concatenate((data_, torch.ones((data_.shape[0], data_.shape[1], 1))), axis=2)
+    transformed_homogeneous = torch.matmul(homogeneous, affine.T)
+    transformed_coordinates = transformed_homogeneous[:, :, :-1]
```

Comment:
```suggestion
    transformed_coordinates = transformed_homogeneous[:, :, :-1] / transformed_homogeneous[:, :, -1:]
```
Ideally the `w` component of the output vectors should be all ones but this isn't always the case and vectors should be scaled by this.

Reply:
Hi @ericspod, I didn't get the point for this update here, this change will scale transformed_coordinates which make output not as expected.

Reply:
I might be wrong here but this is what I thought the math was so someone else please check. For a given affine matrix, if the bottom row isn't $(0\ 0\ 0\ 1)$ then the point you'll get won't necessarily have 1 as the value of its w component. The idea is that the resulting vector should be scaled by that $W$ value to get the correct coordinate. 

If we have some 4x4 affine matrix $A$ and some vector $P$, this vector has to be 1x4 with a $W$ component added which is typically 1:

$$
P = \begin{pmatrix} 
X \\ 
Y \\ 
Z \\ 
W
\end{pmatrix}
$$

Once we apply the matrix multiplication we get $P'$:

$$
P' = AP = \begin{pmatrix} 
X' \\ 
Y' \\ 
Z' \\ 
W'
\end{pmatrix}
$$

This is a 4D vector, so to get the 3D vector we divide by $W'$:

$$
P'_{3D} =\begin{pmatrix} 
X'/W' \\ 
Y'/W' \\ 
Z'/W' 
\end{pmatrix}
$$

If the bottom row of $A$ is $(0\ 0\ 0\ 1)$ then $W'$ will be all 1, in which case it's safe to just remove it, but in the general case the above should be done. 

I'm also surprised how much I can do with Latex here.

Reply:
> I'm also surprised how much I can do with Latex here.

Amazing!

Got it now. I generally assume that w is 1 by default, and I'm uncertain about when it will be different. Moreover, I explicitly adding W in line 2526, so the user doesn’t need to handle it.

Reply:
For what we're doing adding $W$ is fine, though you could check if you're given 4D points and not add it in that case (or if you have 3D points and a 3x3 matrix), but since we're dealing with 2D/3D cartesian points we know adding a dimension of ones is correct.